src/Microsoft.AspNetCore.WebUtilities/KeyValueAccumulator.cs

```diff
@@ -9,24 +9,33 @@ namespace Microsoft.AspNetCore.WebUtilities
 {
     public struct KeyValueAccumulator
     {
-        private Dictionary<string, List<string>> _accumulator;
+        private Dictionary<string, StringValues> _accumulator;
 
         public void Append(string key, string value)
         {
             if (_accumulator == null)
             {
-                _accumulator = new Dictionary<string, List<string>>(StringComparer.OrdinalIgnoreCase);
+                _accumulator = new Dictionary<string, StringValues>(StringComparer.OrdinalIgnoreCase);
             }
-            List<string> values;
+
+            StringValues values;
             if (_accumulator.TryGetValue(key, out values))
             {
-                values.Add(value);
+                if (values.Count <= 1)
+                {
+                    _accumulator[key] = new string[] { values[0], value };
+                }
+                else
+                {
+                    var currentValues = values.ToArray();
+                    Array.Resize(ref currentValues, values.Count + 1);
```

Comment:
There is the checkbox hack double pattern 
```html
<input name="foo" type="checkbox" value="true" />
<input name="foo" type="hidden" value="false" />
```
However this change should still have less allocations than `List` in this case